replication/OrderedReplication.tla

```diff
@@ -0,0 +1,349 @@
+   In order to be able to take Delete, replica needs to be writable, should have a PUT message for the blobid,
+      and shouldn't already have a delete message for the blob in its log *)
+ReplicaCanTakeDelete(blobid, replicaid, orderedMessagesPerReplica, replicaStates1) ==
+    /\ replicaStates1[replicaid] = CanTakeWrite
+    /\ ReplicaHasPutMessage(blobid, replicaid, orderedStoredMessagesPerReplica)
+    \* /\ ~ReplicaHasDeleteMessage(blobid, replicaid, orderedStoredMessagesPerReplica)
+
+ReplicationRequestMessage(sourceReplica, replicaToken) ==
+    [ src |-> sourceReplica, token |-> replicaToken ]
+
+ReplicaRecordAtOffset(replicaid, orderedMessagesPerReplica, offset) ==
+    IF orderedMessagesPerReplica[replicaid] # {} THEN 
+        {msg \in orderedMessagesPerReplica[replicaid]: msg.logoffset = offset}
+    ELSE {}
+
+ReplicationResponseMessage(replicaid, replicaStoredMessages, msg) ==
+    {[r |-> replicaid, messages |-> ReplicaRecordAtOffset(replicaid, replicaStoredMessages, msg.token)]}
+
+TypeOK ==
+    /\ DOMAIN replicaStates \subseteq Replicas
+    /\ DOMAIN inflightReplicationMessages \subseteq Replicas
+    /\ DOMAIN orderedStoredMessagesPerReplica \subseteq Replicas
+    /\ DOMAIN inflightReplicationRequest \subseteq Replicas
+    /\ DOMAIN inflightReplicationResponse \subseteq Replicas
+    /\ DOMAIN replicaTokens \subseteq Replicas
+    /\ DOMAIN replicaOffsets \subseteq Replicas
+    /\ blobIdCtr <= NumBlobs + 1
+
+Init ==
+    /\ orderedStoredMessagesPerReplica = [ r \in Replicas |-> {} ]
+    /\ inflightReplicationMessages = [ r \in Replicas |-> {} ]
+    /\ inflightReplicationRequest = [ r \in Replicas |-> {} ]
+    /\ inflightReplicationResponse = [ r \in Replicas |-> {} ]
+    /\ replicaStates = [ r \in Replicas |-> CanTakeWrite ]
+    /\ blobIdCtr = 0
+    /\ replicaTokens = [r \in Replicas |-> [ rr \in Replicas |-> 0 ] ]
+    /\ replicaOffsets = [r \in Replicas |-> 0 ]
+
+(* If each blob has seen PUTs, updateTtl and Delete then we can assume that all user operations have ended. *)
+AllBlobOpsDone == 
+    /\ blobIdCtr = NumBlobs
+
+(* Take down the replica with the replica id r. *)
+ReplicaDown(r) ==
+    /\ replicaStates' = [replicaStates EXCEPT ![r] = CannotTakeWrite]
+    /\ UNCHANGED <<blobIdCtr, inflightReplicationMessages, inflightReplicationRequest, inflightReplicationResponse, orderedStoredMessagesPerReplica, replicaTokens, replicaOffsets>>
+
+(* Bring up a replica with the replica id r . *)
+ReplicaUp(r) ==
+    /\ replicaStates' = [replicaStates EXCEPT ![r] = CanTakeWrite]
+    /\ UNCHANGED <<blobIdCtr, inflightReplicationMessages, inflightReplicationRequest, inflightReplicationResponse, orderedStoredMessagesPerReplica, replicaTokens, replicaOffsets>>
+
+(* Recieve request to put a blob.
+ *)
+RecvPut ==
+    /\ blobIdCtr <= NumBlobs \* To Remove
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ orderedStoredMessagesPerReplica' = [r \in DOMAIN orderedStoredMessagesPerReplica |-> 
```

Comment:
This looks a little strange to me, but maybe it's because I don't understand it?  It seems like we're setting orderedStoredMessagesPerReplica' to be some element of orderedStoredMessagesPerReplica by setting it equal to orderedStoredMessagesPerReplica[r].  Shouldn't it be something like orderedStoredMessagesPerReplica' = [orderedStoredMessagesPerReplica EXCEPT ....]?

Reply:
oh wait.  I see.  You've effectively done what EXCEPT does with r \in DOMAIN...  Ok, in that case I think this works.

Reply:
Ok, other question.  

\E r1, r2 \in Replicas.  It seems that we do the operation on two Replicas at a time for all these Recv functions.  Why is that?  AFAICT, it looks like we do symmetrical operations for both replicas.

Reply:
We do the put operation on two replicas at a time because thats the quorum for each operation. Each operation is sent to two replicas and is successful only if applied to two replicas.

But you have a good point. There is no guarantee for the operation to be atomic, and the above implementation makes it atomic. Let me try to fix it and update.